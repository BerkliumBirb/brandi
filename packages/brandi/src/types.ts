import { Tag } from './pointers';

type UnknownConstructor<T extends Object = Object> = new (...args: any[]) => T;
type UnknownFunction<T = unknown> = (...args: any[]) => T;

export type UnknownCreator<T = unknown> =
  | UnknownConstructor<T>
  | UnknownFunction<T>;

export type UnknownCreatorParameters<
  T extends UnknownCreator
> = T extends UnknownConstructor
  ? ConstructorParameters<T>
  : T extends UnknownFunction
  ? Parameters<T>
  : never;

export type ResolutionCondition = Tag | UnknownCreator;

export type Factory<T extends unknown, A extends unknown[] = []> = (
  ...args: A
) => T;
