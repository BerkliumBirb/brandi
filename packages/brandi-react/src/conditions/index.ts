export * from './ConditionsContext';
export * from './ConditionsProvider';
export * from './tagged';
export * from './useConditions';
