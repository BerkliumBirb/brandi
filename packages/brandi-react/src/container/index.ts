export * from './ContainerContext';
export * from './ContainerProvider';
export * from './useContainer';
