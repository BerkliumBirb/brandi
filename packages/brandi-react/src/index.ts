export { ContainerProvider } from './container';
export { createInjectionHooks, useInjection } from './injection';
export { tagged } from './tagged';
