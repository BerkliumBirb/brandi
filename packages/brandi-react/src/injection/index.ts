export * from './createInjectionHooks';
export * from './useInjection';
