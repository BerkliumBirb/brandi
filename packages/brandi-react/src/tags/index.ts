export * from './tagged';
export * from './TagsContext';
export * from './TagsProvider';
export * from './useTags';
