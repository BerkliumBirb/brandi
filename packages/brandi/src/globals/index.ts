export * from './entitiesRegistry';
export * from './injectsRegistry';
export * from './tagsRegistry';
