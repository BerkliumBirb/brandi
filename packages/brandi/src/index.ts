export { Container, createContainer } from './container';
export type {
  Tag,
  Token,
  TokenType,
  TokenTypeMap,
  TokenValue,
  RequiredToken,
  OptionalToken,
} from './pointers';
export { tag, token } from './pointers';
export { injected, tagged } from './registrators';
export type { Factory, ResolutionCondition } from './types';
