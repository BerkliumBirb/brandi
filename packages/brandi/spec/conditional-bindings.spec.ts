import { Container, injected, tag, tagged, token } from '../src';

import { setEnv } from './utils';

describe('conditional bindings', () => {
  it('creates an instance with an injection that depends on the target', () => {
    const someValue = 1;
    const anotherValue = 2;

    class SomeClass {
      constructor(public value: number) {}
    }

    class AnotherClass {
      constructor(public value: number) {}
    }

    const tokens = {
      someValue: token<number>('someValue'),
      someClass: token<SomeClass>('someClass'),
      anotherClass: token<AnotherClass>('anotherClass'),
    };

    injected(SomeClass, tokens.someValue);
    injected(AnotherClass, tokens.someValue);

    const container = new Container();

    container.bind(tokens.someValue).toConstant(someValue);
    container
      .when(AnotherClass)
      .bind(tokens.someValue)
      .toConstant(anotherValue);

    container.bind(tokens.someClass).toInstance(SomeClass).inTransientScope();
    container
      .bind(tokens.anotherClass)
      .toInstance(AnotherClass)
      .inTransientScope();

    const someClassInstance = container.get(tokens.someClass);
    const anotherClassInstance = container.get(tokens.anotherClass);

    expect(someClassInstance.value).toBe(someValue);
    expect(anotherClassInstance.value).toBe(anotherValue);
  });

  it('creates an instance with an injection that depends on the tag', () => {
    const someValue = 1;
    const anotherValue = 2;

    class SomeClass {
      constructor(public value: number) {}
    }

    class AnotherClass {
      constructor(public value: number) {}
    }

    const tokens = {
      someValue: token<number>('someValue'),
      someClass: token<SomeClass>('someClass'),
      anotherClass: token<AnotherClass>('anotherClass'),
    };

    const tags = {
      some: tag('some'),
    };

    injected(SomeClass, tokens.someValue);

    injected(AnotherClass, tokens.someValue);
    tagged(AnotherClass, tags.some);

    const container = new Container();

    container.bind(tokens.someValue).toConstant(someValue);
    container.when(tags.some).bind(tokens.someValue).toConstant(anotherValue);

    container.bind(tokens.someClass).toInstance(SomeClass).inTransientScope();
    container
      .bind(tokens.anotherClass)
      .toInstance(AnotherClass)
      .inTransientScope();

    const someClassInstance = container.get(tokens.someClass);
    const anotherClassInstance = container.get(tokens.anotherClass);

    expect(someClassInstance.value).toBe(someValue);
    expect(anotherClassInstance.value).toBe(anotherValue);
  });

  it('injects a dependency by the target condition when there are conditions for both the target and the tag', () => {
    const someValue = 1;
    const anotherValue = 2;

    class SomeClass {
      constructor(public value: number) {}
    }

    class AnotherClass {
      constructor(public value: number) {}
    }

    const tokens = {
      someValue: token<number>('someValue'),
      someClass: token<SomeClass>('someClass'),
      anotherClass: token<AnotherClass>('anotherClass'),
    };

    const tags = {
      some: tag('some'),
    };

    injected(SomeClass, tokens.someValue);
    tagged(SomeClass, tags.some);

    injected(AnotherClass, tokens.someValue);
    tagged(AnotherClass, tags.some);

    const container = new Container();

    container.when(tags.some).bind(tokens.someValue).toConstant(someValue);
    container
      .when(AnotherClass)
      .bind(tokens.someValue)
      .toConstant(anotherValue);

    container.bind(tokens.someClass).toInstance(SomeClass).inTransientScope();
    container
      .bind(tokens.anotherClass)
      .toInstance(AnotherClass)
      .inTransientScope();

    const someClassInstance = container.get(tokens.someClass);
    const anotherClassInstance = container.get(tokens.anotherClass);

    expect(someClassInstance.value).toBe(someValue);
    expect(anotherClassInstance.value).toBe(anotherValue);
  });

  it('ignores an unused tag on the target', () => {
    const someValue = 1;

    class SomeClass {
      constructor(public value: number) {}
    }

    const tokens = {
      someValue: token<number>('someValue'),
      someClass: token<SomeClass>('someClass'),
    };

    const tags = {
      unused: tag('unused'),
    };

    injected(SomeClass, tokens.someValue);

    tagged(SomeClass, tags.unused);

    const container = new Container();
    container.bind(tokens.someValue).toConstant(someValue);
    container.bind(tokens.someClass).toInstance(SomeClass).inTransientScope();

    const instance = container.get(tokens.someClass);

    expect(instance.value).toBe(someValue);
  });

  it('injects a dependency by conditions from the parent container', () => {
    const defaultValue = 1;
    const someValue = 2;
    const anotherValue = 3;

    class SomeClass {
      constructor(public value: number) {}
    }
    class AnotherClass {
      constructor(public value: number) {}
    }

    const tokens = {
      value: token<number>('value'),
      someClass: token<SomeClass>('someClass'),
      anotherClass: token<AnotherClass>('anotherClass'),
    };

    const tags = {
      some: tag('some'),
    };

    injected(SomeClass, tokens.value);
    tagged(SomeClass, tags.some);

    injected(AnotherClass, tokens.value);

    const parentContainer = new Container();

    parentContainer.bind(tokens.value).toConstant(defaultValue);
    parentContainer.when(tags.some).bind(tokens.value).toConstant(someValue);
    parentContainer
      .when(AnotherClass)
      .bind(tokens.value)
      .toConstant(anotherValue);

    const childContainer = new Container(parentContainer);
    childContainer
      .bind(tokens.someClass)
      .toInstance(SomeClass)
      .inTransientScope();
    childContainer
      .bind(tokens.anotherClass)
      .toInstance(AnotherClass)
      .inTransientScope();

    const someInstance = childContainer.get(tokens.someClass);
    const anotherInstance = childContainer.get(tokens.anotherClass);

    expect(someInstance.value).toBe(someValue);
    expect(anotherInstance.value).toBe(anotherValue);
  });

  it('takes a binding by the tag assigned first with multiple related tags on the target', () => {
    const spy = jest.spyOn(console, 'warn').mockImplementation(() => null);

    const someValue = 1;
    const anotherValue = 2;
    const otherValue = 3;

    class SomeClass {
      constructor(public value: number) {}
    }

    const tokens = {
      someValue: token<number>('someValue'),
      someClass: token<SomeClass>('someClass'),
    };

    const tags = {
      some: tag('some'),
      another: tag('another'),
      other: tag('other'),
    };

    injected(SomeClass, tokens.someValue);
    tagged(SomeClass, tags.some, tags.another, tags.other);

    const container = new Container();

    container.when(tags.other).bind(tokens.someValue).toConstant(otherValue);
    container.when(tags.some).bind(tokens.someValue).toConstant(someValue);
    container
      .when(tags.another)
      .bind(tokens.someValue)
      .toConstant(anotherValue);

    container.bind(tokens.someClass).toInstance(SomeClass).inTransientScope();

    const instance = container.get(tokens.someClass);

    expect(instance.value).toBe(someValue);

    spy.mockRestore();
  });

  it('logs a warning when multiple related tags are on the target', () => {
    const spy = jest.spyOn(console, 'warn').mockImplementation(() => null);

    class SomeClass {
      constructor(public value: number) {}
    }

    const tokens = {
      someValue: token<number>('someValue'),
      someClass: token<SomeClass>('someClass'),
    };

    const tags = {
      some: tag('some'),
      another: tag('another'),
    };

    injected(SomeClass, tokens.someValue);
    tagged(SomeClass, tags.some, tags.another);

    const container = new Container();

    container.when(tags.another).bind(tokens.someValue).toConstant(0);
    container.when(tags.some).bind(tokens.someValue).toConstant(0);

    container.bind(tokens.someClass).toInstance(SomeClass).inTransientScope();

    container.get(tokens.someClass);

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy.mock.calls[0]?.[0]).toMatchSnapshot();

    spy.mockRestore();
  });

  it('does not log a warning when a single related tag are on the target', () => {
    const spy = jest.spyOn(console, 'warn').mockImplementation(() => null);

    class SomeClass {
      constructor(public value: number) {}
    }

    class AnotherClass {
      constructor(public value: number) {}
    }

    const tokens = {
      someValue: token<number>('someValue'),
      someClass: token<SomeClass>('someClass'),
      anotherClass: token<AnotherClass>('anotherClass'),
    };

    const tags = {
      some: tag('some'),
    };

    injected(SomeClass, tokens.someValue);
    injected(AnotherClass, tokens.someValue);

    tagged(AnotherClass, tags.some);

    const container = new Container();

    container.bind(tokens.someValue).toConstant(0);
    container.when(tags.some).bind(tokens.someValue).toConstant(0);

    container.bind(tokens.someClass).toInstance(SomeClass).inTransientScope();
    container
      .bind(tokens.anotherClass)
      .toInstance(AnotherClass)
      .inTransientScope();

    container.get(tokens.someClass);
    container.get(tokens.anotherClass);

    expect(spy).toHaveBeenCalledTimes(0);

    spy.mockRestore();
  });

  it("skips the logging in 'production' env", () => {
    const restoreEnv = setEnv('production');
    const spy = jest.spyOn(console, 'warn').mockImplementation(() => null);

    class SomeClass {
      constructor(public value: number) {}
    }

    const tokens = {
      someValue: token<number>('someValue'),
      someClass: token<SomeClass>('someClass'),
    };

    const tags = {
      some: tag('some'),
      another: tag('another'),
    };

    injected(SomeClass, tokens.someValue);
    tagged(SomeClass, tags.some, tags.another);

    const container = new Container();

    container.when(tags.another).bind(tokens.someValue).toConstant(0);
    container.when(tags.some).bind(tokens.someValue).toConstant(0);

    container.bind(tokens.someClass).toInstance(SomeClass).inTransientScope();

    container.get(tokens.someClass);

    expect(spy).toHaveBeenCalledTimes(0);

    restoreEnv();
    spy.mockRestore();
  });

  describe('Container.get', () => {
    it('returns a dependency with a condition', () => {
      const spy = jest.spyOn(console, 'warn').mockImplementation(() => null);

      const someValue = 1;
      const anotherValue = 2;
      const otherValue = 3;

      class SomeClass {
        constructor(public value: number) {}
      }

      const tokens = {
        someValue: token<number>('someValue'),
      };

      const tags = {
        some: tag('some'),
      };

      injected(SomeClass, tokens.someValue);

      const container = new Container();

      container.bind(tokens.someValue).toConstant(someValue);
      container.when(tags.some).bind(tokens.someValue).toConstant(anotherValue);
      container.when(SomeClass).bind(tokens.someValue).toConstant(otherValue);

      expect(container.get(tokens.someValue)).toBe(someValue);
      expect(container.get(tokens.someValue, [])).toBe(someValue);
      expect(container.get(tokens.someValue, [tags.some])).toBe(anotherValue);
      expect(container.get(tokens.someValue, [SomeClass])).toBe(otherValue);
      expect(container.get(tokens.someValue, [tags.some, SomeClass])).toBe(
        anotherValue,
      );
      expect(container.get(tokens.someValue, [SomeClass, tags.some])).toBe(
        otherValue,
      );

      spy.mockRestore();
    });

    it('logs a warning when multiple related conditions were passed', () => {
      const spy = jest.spyOn(console, 'warn').mockImplementation(() => null);

      class SomeClass {
        constructor(public value: number) {}
      }

      const tokens = {
        someValue: token<number>('someValue'),
      };

      const tags = {
        some: tag('some'),
      };

      injected(SomeClass, tokens.someValue);

      const container = new Container();

      container.bind(tokens.someValue).toConstant(0);
      container.when(tags.some).bind(tokens.someValue).toConstant(0);
      container.when(SomeClass).bind(tokens.someValue).toConstant(0);

      container.get(tokens.someValue, [SomeClass, tags.some]);

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy.mock.calls[0]?.[0]).toMatchSnapshot();

      spy.mockRestore();
    });

    it('does not log a warning when a single related condition were passed', () => {
      const spy = jest.spyOn(console, 'warn').mockImplementation(() => null);

      class SomeClass {
        constructor(public value: number) {}
      }

      const tokens = {
        someValue: token<number>('someValue'),
      };

      const tags = {
        some: tag('some'),
      };

      injected(SomeClass, tokens.someValue);

      const container = new Container();

      container.bind(tokens.someValue).toConstant(0);
      container.when(tags.some).bind(tokens.someValue).toConstant(0);
      container.when(SomeClass).bind(tokens.someValue).toConstant(0);

      container.get(tokens.someValue, [SomeClass]);
      container.get(tokens.someValue, [tags.some]);

      expect(spy).toHaveBeenCalledTimes(0);

      spy.mockRestore();
    });

    it("skips the logging in 'production' env", () => {
      const restoreEnv = setEnv('production');
      const spy = jest.spyOn(console, 'warn').mockImplementation(() => null);

      class SomeClass {
        constructor(public value: number) {}
      }

      const tokens = {
        someValue: token<number>('someValue'),
      };

      const tags = {
        some: tag('some'),
      };

      injected(SomeClass, tokens.someValue);

      const container = new Container();

      container.bind(tokens.someValue).toConstant(0);
      container.when(tags.some).bind(tokens.someValue).toConstant(0);
      container.when(SomeClass).bind(tokens.someValue).toConstant(0);

      container.get(tokens.someValue, [SomeClass, tags.some]);

      expect(spy).toHaveBeenCalledTimes(0);

      restoreEnv();
      spy.mockRestore();
    });
  });
});
