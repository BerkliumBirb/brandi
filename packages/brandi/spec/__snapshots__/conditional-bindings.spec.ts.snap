// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conditional bindings Container.get logs a warning when multiple related conditions were passed 1`] = `"Warning: When resolving a binding by 'someValue' token with [SomeClass, tag(some)] conditions, more than one binding was found. In this case, Brandi resolves the binding by the first tag assigned by 'tagged(target, ...tags)' function or, if you explicitly passed conditions through 'Container.get(token, conditions)' method, by the first resolved condition. Try to avoid such implicit logic."`;

exports[`conditional bindings logs a warning when multiple related tags are on the target 1`] = `"Warning: When resolving a binding by 'someValue' token with [tag(some), tag(another)] conditions, more than one binding was found. In this case, Brandi resolves the binding by the first tag assigned by 'tagged(target, ...tags)' function or, if you explicitly passed conditions through 'Container.get(token, conditions)' method, by the first resolved condition. Try to avoid such implicit logic."`;
